### DISCUSSION TITLE
Custom Compilation of BooleanClauseList to remove Parenthesis surrounding EXISTS

### DISCUSSION THREAD

**Discussion by mkmoisen** [sqlalchemy/sqlalchemy#12070] [Usage Questions]
Hi @zzzeek 

Would you please review my attempt at `@compiles(BooleanClauseList)` ?

SQLAlchemy by default will wrap an EXISTS in parenthesis if used in the WHERE clause with other conditions separated by AND/OR. It does not wrap the exists with parenthesis if the EXISTS is the only condition in the where clause.

I would like to have a custom compilation for `BooleanClauseList` which eliminates the wrapped parenthesis.

With ChatGPT's help I've narrowed it down to the following:

```
@compiles(BooleanClauseList)
def compile_boolean_clause_list(element, compiler, **kw):
    compiled_clauses = []

    for clause in element.clauses:
        # EXISTS expressions are inside of Grouping
        if isinstance(clause, Grouping):
            inner = clause.element
            while isinstance(inner, Grouping):
                inner = inner.element

            if isinstance(inner, Exists):
                # Sets the clause to the EXISTS instead of the ( EXISTS )
                clause = inner

        compiled_clauses.append(
            compiler.process(clause, **kw)
        )

    operator = " AND "
    if element.operator is operators.or_:
        operator = " OR "

    return operator.join(compiled_clauses)
```
    
This appears to be working but I have a few concerns.

Are there any other operators I need to worry about, aside from AND and OR?

Is there any reason that SQLAlchemy is wrapping EXISTS with parenthesis in the first place? I'm under the impression this is not required.

Do you think this function is safe in general? I would hate for it to result in bugs elsewhere.

Thanks!

**Comment by zzzeek** [comment]
hi -

I like to work from discussion of problems before we seek writing code to change things.

What's wrong with the way parenthesis are used with EXISTS ?

to assist you in writing code, please work with the below program which illustrates EXISTS used in WHERE both with and without an AND.  The results are correct. Can you modify this program to show an incorrect result due to the parenthesis?

```py
from __future__ import annotations

from sqlalchemy import create_engine
from sqlalchemy import exists
from sqlalchemy import ForeignKey
from sqlalchemy import select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str]
    bs: Mapped[list[B]] = relationship("B")


class B(Base):
    __tablename__ = "b"
    id: Mapped[int] = mapped_column(primary_key=True)
    a_id: Mapped[int] = mapped_column(ForeignKey("a.id"))
    data: Mapped[str]


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add_all(
    [
        A(data="d1", bs=[B(data="db1")]),
        A(data="d2", bs=[B(data="db2")]),
        A(data="d3", bs=[B(data="db2"), B(data="db3")]),
    ]
)
s.commit()


stmt = select(A).where(exists().where(B.a_id == A.id, B.data == "db2"))

print(s.scalars(stmt).all())


stmt = select(A).where(
    exists().where(B.a_id == A.id, B.data == "db2"),
    exists().where(B.a_id == A.id, B.data == "db3"),
)

print(s.scalars(stmt).all())

```

**Comment by mkmoisen** [comment]
Hi @zzzeek 

Nothing is wrong with wrapping the EXISTS with parenthesis, except I simply don't like the way it looks when I format the SQL.

We use SQLAlchemy core to automate the generation of Oracle PL/SQL packages, and we format the generated SQL with sqlglot such that the SQL corresponds to the company's formatting standards.

So with EXISTS wrapped in parenthesis the result is something like:

```
SELECT foo
FROM bar
WHERE 1=1
    AND (
        EXISTS (
            SELECT 1
            FROM ...
        )
    )
```

And I just have a visual preference for:

```
SELECT foo
FROM bar
WHERE 1=1
    AND EXISTS (
        SELECT 1
        FROM ...
    )
```

That is the sole reason for the desire to remove the parenthesis surrounding EXISTS.

Thank you.

**Reply by zzzeek** [reply]
well to your original question I have no idea if the code you have presented might have edge cases that cause invalid SQL, the nesting parenthesis issue gets pretty thorny with some kinds of edge cases, then again as it seems your logic is just to flatten additional groupings, maybe it's fine and on the SQLAlchemy side is only something we dont have because it adds extra complexity to the compilation stage.  If it works for your immediate needs and nothing is breaking, I would advise writing appropriate unit tests for everything you need to do and then you should be fine.
